crates/environ/src/vmoffsets.rs

```diff
@@ -181,103 +182,63 @@ impl<P: PtrSize> From<VMOffsetsFields<P>> for VMOffsets<P> {
-                    .unwrap(),
-            )
-            .unwrap();
-        ret.defined_tables = ret
-            .imported_globals
-            .checked_add(
-                ret.num_imported_globals
-                    .checked_mul(u32::from(ret.size_of_vmglobal_import()))
-                    .unwrap(),
-            )
-            .unwrap();
-        ret.defined_memories = ret
-            .defined_tables
-            .checked_add(
-                ret.num_defined_tables
-                    .checked_mul(u32::from(ret.size_of_vmtable_definition()))
-                    .unwrap(),
-            )
-            .unwrap();
-        ret.defined_globals = align(
-            ret.defined_memories
-                .checked_add(
-                    ret.num_defined_memories
-                        .checked_mul(u32::from(ret.size_of_vmmemory_definition()))
-                        .unwrap(),
-                )
-                .unwrap(),
-            16,
-        );
-        ret.defined_anyfuncs = ret
-            .defined_globals
-            .checked_add(
-                ret.num_defined_globals
-                    .checked_mul(u32::from(ret.size_of_vmglobal_definition()))
-                    .unwrap(),
-            )
-            .unwrap();
-        ret.builtin_functions = ret
-            .defined_anyfuncs
-            .checked_add(
-                ret.num_imported_functions
-                    .checked_add(ret.num_defined_functions)
-                    .unwrap()
-                    .checked_mul(u32::from(ret.size_of_vmcaller_checked_anyfunc()))
-                    .unwrap(),
-            )
-            .unwrap();
-        ret.size = ret
-            .builtin_functions
-            .checked_add(u32::from(ret.pointer_size()))
-            .unwrap();
+        // Use a local variable instead of ret.size to prevent unnecessary loads and stores.
+        // LLVM is not able to optimize accesses of ret.size away most of the time.
+        let mut next_field_offset = 0;
+
+        // Convenience functions for checked addition and multiplication.
+        // As side effect this reduces binary size by using only a single
+        // `#[track_caller]` location for each function instead of one for
+        // each individual invocation.
+        #[inline]
+        fn cadd(count: u32, size: u32) -> u32 {
+            count.checked_add(size).unwrap()
+        }
+
+        #[inline]
+        fn cmul(count: u32, size: u8) -> u32 {
+            count.checked_mul(u32::from(size)).unwrap()
+        }
+
+        macro_rules! field {
+            ($field:ident = .; size = $size:expr) => {
+                ret.$field = next_field_offset;
+                next_field_offset = cadd(next_field_offset, $size);
+            };
+            ($field:ident = align(., $align:literal); size = $size:expr) => {
+                ret.$field = align(next_field_offset, $align);
+                next_field_offset = cadd(ret.$field, $size);
+            };
+        }
```

Comment:
Personally I find this macro pretty hard to read and the syntax feels pretty odd. Could this perhaps be improved with something like:

```rust
fields! {
    size(interrupts) = ret.ptr.size(),
    size(epoch_ptr) = ret.ptr.size(),
    // ...
}
```

I think it might be helpful for the macro to automatically insert `u32::from` everywhere to avoid having to explicitly call it as well. Furthermore I think it's best to avoid `cmul` and `cadd` aliases and instead just have the `.checked_mul` inline since at least for me I think that's easier to read.

Reply:
> Personally I find this macro pretty hard to read and the syntax feels pretty odd.

I guess so. I like your suggestion.

> Furthermore I think it's best to avoid cmul and cadd aliases and instead just have the .checked_mul inline since at least for me I think that's easier to read.

`a.checked_mul(b).unwrap()` is harder to read than `cmul(a, b)` IMHO. Especially as the former often requires wrapping to stay below 100 characters. Also as noted in the comment above their definition it increases binary size a bit due to unique caller locations passed to `.unwrap()`. If you aren't convinced I will drop it.

Reply:
> Personally I find this macro pretty hard to read and the syntax feels pretty odd. Could this perhaps be improved with something like:

How should align be handled? `defined_globals` needs to be aligned to 16 bytes.

Reply:
I'm personally not a huge fan of one-off functions with short names like "cmul" purely to save some characters on a few lines. Others may feel differently.

For alignment I would naively think that `align(the_field) = align` could be a macro argument which simply aligns up the current offset and implicitly the alignment is always specified after the size. (or before the size? something like that)

Reply:
Maybe something like

```rust
fields! {
    size(interrupts) = ret.ptr.size(),
    size(epoch_ptr) = ret.ptr.size(),
    // ...
    align(16),
    size(defined_globals) = ...,
    // ...
}
```